packages/datadog-plugin-aws-sdk/src/index.js

```diff
@@ -0,0 +1,109 @@
+'use strict'
+
+const Tags = require('opentracing').Tags
+const analyticsSampler = require('../../dd-trace/src/analytics_sampler')
+const awsHelpers = require('./helpers')
+
+function createWrapRequest (tracer, config) {
+  config = normalizeConfig(config)
+
+  return function wrapRequest (request) {
+    return function requestWithTrace (operation, params, cb) {
+      const serviceName = awsHelpers.normalizeServiceName(this)
+      const childOf = tracer.scope().active()
+      const baseTags = {
+        [Tags.SPAN_KIND]: 'client',
+        'span.type': 'http',
+        'service.name': config.service ? `${config.service}-${serviceName}` : `${tracer._service}-${serviceName}`,
+        'aws.agent': 'js-aws-sdk',
+        'aws.operation': operation,
+        'aws.region': this.config.region,
+        'aws.service': serviceName,
+        'aws.url': this.endpoint && this.endpoint.href,
+        'component': serviceName
+      }
+      let span
+
+      if (typeof cb === 'function') {
+        span = tracer.startSpan('aws.http', {
+          childOf,
+          tags: baseTags
+        })
+
+        analyticsSampler.sample(span, config.analytics)
+
+        // when passed a callback makeRequest will call `.send` on the AWS.Request
+        // before we have a chance to add our listeners, so we have to wrap
+        // the callback in order know when the span has finished
+        // https://github.com/aws/aws-sdk-js/blob/
+        // 9c191bbdbf32a8a3fa31219e369006f852318a1f/lib/service.js#L202-L206
+        return tracer.scope().activate(span, () => {
+          return request.call(
+            this,
+            operation,
+            params,
+            awsHelpers.wrapCallback(tracer, span, cb, childOf, config, serviceName)
+          )
+        })
+      } else {
+        const awsReq = request.apply(this, arguments)
+        const boundAwsReq = tracer.scope().bind(awsReq, childOf)
+
+        // The state machine has garaunteed first and last events, validate / complete
+        // instrumenting those ensures requests w/out callback gets started/ended correctly
+        // https://github.com/aws/aws-sdk-js/blob/38bf84c144281f696768e8c64500f2847fe6f298/lib/request.js#L142
+        // https://github.com/aws/aws-sdk-js/blob/38bf84c144281f696768e8c64500f2847fe6f298/lib/request.js#L328-L332
+        boundAwsReq.on('validate', () => {
+          if (span) return
+
+          span = tracer.startSpan('aws.http', {
+            childOf,
+            tags: baseTags
+          })
+
+          analyticsSampler.sample(span, config.analytics)
+          tracer.scope().activate(span)
+        })
+
+        boundAwsReq.on('complete', response => {
```

Comment:
Can this fail? How is the failure reported?

Reply:
afaik, this is what the state machine looks like

```
  this.addState('validate', 'build', 'error', transition);
  this.addState('build', 'afterBuild', 'restart', transition);
  this.addState('afterBuild', 'sign', 'restart', transition);
  this.addState('sign', 'send', 'retry', transition);
  this.addState('retry', 'afterRetry', 'afterRetry', transition);
  this.addState('afterRetry', 'sign', 'error', transition);
  this.addState('send', 'validateResponse', 'retry', transition);
  this.addState('validateResponse', 'extractData', 'extractError', transition);
  this.addState('extractError', 'extractData', 'retry', transition);
  this.addState('extractData', 'success', 'retry', transition);
  this.addState('restart', 'build', 'error', transition);
  this.addState('success', 'complete', 'complete', transition);
  this.addState('error', 'complete', 'complete', transition);
  this.addState('complete', null, null, transition);
```

  - from: https://github.com/aws/aws-sdk-js/blob/38bf84c144281f696768e8c64500f2847fe6f298/lib/request.js#L44-L57

with `addState` being a function that accept the state, the success state, and the fail state:

  - `function addState(name, acceptState, failState, fn)`
  - from: https://github.com/aws/aws-sdk-js/blob/cc29728c1c4178969ebabe3bbe6b6f3159436394/lib/state_machine.js#L30

At each `transition` the new state is emitted

  - `self.emit(self._asm.currentState, ...`

  - from: https://github.com/aws/aws-sdk-js/blob/38bf84c144281f696768e8c64500f2847fe6f298/lib/request.js#L22

Which is what triggers our instrumentation's `validate` and `complete` callbacks

In the case `validate` fails, it would transition to `error`, which then transitions to `complete`. 

```
this.addState('validate', 'build', 'error', transition);
...
this.addState('error', 'complete', 'complete', transition);
```

`complete` receives a `response` object which includes `data` and `error` keys, this `response.error` key is what we pass into `.finish` and in `.finish` we check for the existence of any errors.

Basically all errors or successes within the state machine eventually get funneled to `complete` no matter what, and `complete` is the end of the request cycle.

As I understand it and from my tests, this means we should be correctly accounting for any errors, and not interfering with how they're handled.